### ISSUE TITLE
Promise error type

### ISSUE THREAD

**Issue by LoganBarnett** [facebook/flow#5042]
The Promise type starts its annotation like so:
``` javascript
declare class Promise<+R> {
  ...
```

For errors, it has:
``` javascript
    catch<U>(
      onReject?: (error: any) => Promise<U> | U
    ): Promise<R | U>;
```

Is this basically asserting we can't possibly type a promise's error? I would think I could provide something akin to `Error | RejectUnionInMyApp`, especially if I control the promise chain from top to bottom.

For now I can setup my callbacks to be annotated with `(e: E) => ...` but it does feel like a bandaid. Thoughts?

**Comment by TiddoLangerak** [comment]
For this to work flow first need to get support for typed exceptions: a promise is not only rejected when `reject` is called explicitly, but also when an error is thrown. Since flow doesn't know what errors can be thrown, it also cannot type `onReject` any stronger. 

As an example. suppose promise would get an error type as second parameter:

```js
function stringThrower() : string {
  throw "Some string";
}

// We say to flow that the promise only rejects with an Error.
const myPromise : Promise<string, Error> = new Promise((resolve, reject) => {
  // but in reality `stringThrower` will cause the promise to be rejected with a string. 
  // There's currently no way for flow to detect this, so flow would happily accept this. 
  resolve(stringThrower());
});

// And then this becomes unsound:
myPromise.catch((error : Error) => {
  const stack = error.stack.split('\n');
});
```

**Comment by LoganBarnett** [comment]
Thanks for getting back to me!

I can see why that would cause problems for being sound. It would be nice if the type were `mixed` instead of `any`, but there's another ticket for that particular ask for the type of the `catch` variable (#2470), and that seems to be a prerequisite for this.

I'm good closing this or tagging it as blocked on the other issue - whatever folks prefer.

**Comment by Gozala** [comment]
I have started an effort to add an error type parameter to a promise class #5634

**Comment by SamChou19815** [comment]
We are unlikely to pursue this to unnecessarily deviate from TS.
